framework/include/cppmicroservices/AnyMap.h

```diff
@@ -34,13 +34,27 @@ namespace detail {
 
 struct any_map_cihash
 {
-  std::size_t operator()(const std::string& key) const;
+  std::size_t operator()(const std::string& key) const
```

Comment:
Is there a good reason for moving the definitions to the header file?

Reply:
Moved inline because I added a test that checks a couple of simple things about the hash, and that required that the symbol either be exported from the cpp file, or that the hash algorithm be inline. I found it easier to deal with inline, so I did that. If that's a problem I can make the change to export the symbol instead.

Reply:
I am not a proponent of exporting symbols or putting implementation details (definitions) in headers to support tests (unless there is a really good reason). In this case, I think you could test the hash function implicitly by using a any_map with special test values.

Reply:
Ordinarily I'd agree. However, I think it's important that we test the hash algorithm (which was not being tested before) itself, and not the use of the algorithm in the code. That is, we should treat the hash algorithm as the unit under test and test it directly. Also, The hash algorithm is in the header file for AnyMap which has implementation details (because of all the templated code). Revealing the particulars of the cihash in comparison to other details that are visible there seems to be of minor concern.

So, in sum, I think it's important that we test the hash algorithm, and I don't think it's worth a lot of code reorganization for the very few lines of implementation code exposed.

Reply:
In this particular case, I disagree. There is no code reorganization required (actually, the proposed changes need more reorganization - moving the implementation to the header). The other details which are already exposed are out of necessity (if not, I would advocate to move them as well).

The simple tests for hash equality and non-equality are equivalent to testing if the unordered_map contains the case-variant key or not. This *is* testing the hash algorithm.

If you want to test the hash class directly, why not use `std::unordered_map<...>::hasher` instead of exposing its details?

Reply:
I just tried using any_map::unordered_any_cimap::hasher in the tests and that gives me an unresolved symbol because the the operator() isn't exported. I can add the US_Framework_EXPORT macro to the any_map_cihash and any_map_ciequal structs and that fixes the symbol problem. Would that be preferable to you? If so, I can submit that change.

As far as testing the map for whether or not it contains the key, while this has the effect of testing the hash algorithm, that's testing it as a side effect. I think it's better in general to actually write unit tests to exercise exactly the code under test rather than trying to do so by manipulating the calling code.

Reply:
Thanks for trying `hasher`, adding the export macro is okay for me.

I agree in general with what you say. In our case, it affects the library API surface, potential backwards compatibility, (I know we do not provide strong guarantees for that), compile dependencies, etc. But I think we found a good compromise here.

Reply:
OK. I just pushed these changes in. Once the ci builds finish we can pull the changes into development. Thanks for taking the time to review :-)